app/controllers/api/messaging.php

```diff
@@ -0,0 +1,2145 @@
+            throw new Exception(Exception::SUBSCRIBER_NOT_FOUND);
+        }
+
+        $response
+            ->dynamic($subscriber, Response::MODEL_SUBSCRIBER);
+    });
+
+App::delete('/v1/messaging/topics/:topicId/subscriber/:subscriberId')
+    ->desc('Delete a subscriber from a topic.')
+    ->groups(['api', 'messaging'])
+    ->label('audits.event', 'subscriber.delete')
+    ->label('audits.resource', 'subscriber/{request.subscriberId}')
+    ->label('scope', 'subscribers.write')
+    ->label('sdk.auth', [APP_AUTH_TYPE_JWT, APP_AUTH_TYPE_SESSION, APP_AUTH_TYPE_ADMIN, APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'messaging')
+    ->label('sdk.method', 'deleteSubscriber')
+    ->label('sdk.description', '/docs/references/messaging/delete-subscriber.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_NOCONTENT)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_NONE)
+    ->param('topicId', '', new UID(), 'Topic ID.')
+    ->param('subscriberId', '', new UID(), 'Subscriber ID.')
+    ->inject('dbForProject')
+    ->inject('response')
+    ->action(function (string $topicId, string $subscriberId, Database $dbForProject, Response $response) {
+        $topic = Authorization::skip(fn () => $dbForProject->getDocument('topics', $topicId));
+
+        if ($topic->isEmpty()) {
+            throw new Exception(Exception::TOPIC_NOT_FOUND);
+        }
+
+        $subscriber = $dbForProject->getDocument('subscribers', $subscriberId);
+
+        if ($subscriber->isEmpty() || $subscriber->getAttribute('topicId') !== $topicId) {
+            throw new Exception(Exception::SUBSCRIBER_NOT_FOUND);
+        }
+
+        $subscriber = $dbForProject->deleteDocument('subscribers', $subscriberId);
+        Authorization::skip(fn () => $dbForProject->decreaseDocumentAttribute('topics', $topicId, 'total', 1));
+
+        $response
+            ->setStatusCode(Response::STATUS_CODE_NOCONTENT)
+            ->noContent();
+    });
+
+App::post('/v1/messaging/messages/email')
+    ->desc('Create an email.')
+    ->groups(['api', 'messaging'])
+    ->label('audits.event', 'message.create')
+    ->label('audits.resource', 'message/{response.$id}')
+    ->label('scope', 'messages.write')
+    ->label('sdk.auth', [APP_AUTH_TYPE_ADMIN, APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'messaging')
+    ->label('sdk.method', 'createEmail')
+    ->label('sdk.description', '/docs/references/messaging/create-email.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_CREATED)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_MESSAGE)
+    ->param('messageId', '', new CustomId(), 'Message ID. Choose a custom ID or generate a random ID with `ID.unique()`. Valid chars are a-z, A-Z, 0-9, period, hyphen, and underscore. Can\'t start with a special char. Max length is 36 chars.')
```

Comment:
shouldn't these create email/sms/push also take a providerId as input?

Reply:
Jake said:

> Not anymore, we changed this so that topics can support multiple providers. One API call may actually send 3 messages to different providers, so the message needs to be agnostic to provider. In the worker the targets are grouped by provider and a message is sent to each